pyiceberg/io/pyarrow.py

```diff
@@ -1249,11 +1251,12 @@ def _task_to_record_batches(
                     # https://github.com/apache/arrow/issues/39220
                     arrow_table = pa.Table.from_batches([batch])
                     arrow_table = arrow_table.filter(pyarrow_filter)
+                    if len(arrow_table) == 0:
+                        continue
                     batch = arrow_table.to_batches()[0]
             yield _to_requested_schema(
                 projected_schema, file_project_schema, batch, downcast_ns_timestamp_to_us=True, use_large_types=use_large_types
             )
-            current_index += len(batch)
```

Comment:
When working on fixing https://github.com/apache/iceberg-python/issues/1024 I realized a correctness issue was introduced here because we are using the length of the filtered batch instead of the original one when tracking the `current_index`. I think it'll be crucial to get this fix in with 0.7.1 as soon as possible to support our MOR users

Reply:
Oof, that's a good find. Thanks @vhnguyenae for reporting this!

The order of applying filters also caught me when implementing positional deletes. In the long run, I think it would be good to push this down to Arrow, I created an issue a while ago: https://github.com/apache/arrow/issues/35301 But that hasn't seen much traction.